src/Controller/GoogleAuthController.php

```diff
@@ -74,45 +111,134 @@ public static function create(ContainerInterface $container) {
+    /* @var \League\OAuth2\Client\Provider\Google false $google */
+    $google = $this->networkManager->createInstance('social_auth_google')->getSdk();
+
+    // If google client could not be obtained.
+    if (!$google) {
+      drupal_set_message($this->t('Social Auth Google not configured properly. Contact site administrator.'), 'error');
+      return $this->redirect('user.login');
+    }
+
+    // Google service was returned, inject it to $googleManager.
+    $this->googleManager->setClient($google);
+
+    $data_points = explode(',', $this->getDataPoints());
 
-    return new RedirectResponse($client->createAuthUrl());
+    // Generates the URL where the user will be redirected for Google login.
+    // If the user did not have email permission granted on previous attempt,
+    // we use the re-request URL requesting only the email address.
+    $google_login_url = $this->googleManager->getGoogleLoginUrl($data_points);
+
+    $state = $this->googleManager->getState();
+
+    $this->dataHandler->set('oAuth2State', $state);
+
+    return new TrustedRedirectResponse($google_login_url);
   }
 
   /**
-   * Callback function to login user.
+   * Response for path 'user/login/google/callback'.
+   *
+   * Google returns the user here after user has authenticated in Google.
    */
   public function callback() {
-    /* @var \Google_Client $client */
-    $client = $this->networkManager->createInstance('social_auth_google')->getSdk();
+    // Checks if user cancel login via Google.
+    $error = $this->request->getCurrentRequest()->get('error');
+    if ($error == 'access_denied') {
+      drupal_set_message($this->t('You could not be authenticated.'), 'error');
+      return $this->redirect('user.login');
+    }
 
-    $this->googleManager->setClient($client)->authenticate();
+    /* @var \League\OAuth2\Client\Provider\Google false $google */
+    $google = $this->networkManager->createInstance('social_auth_google')->getSdk();
 
-    // Saves access token so that event subscribers can call Google API.
-    $this->session->set('social_auth_google_access_token', $this->googleManager->getAccessToken());
+    // If Google client could not be obtained.
+    if (!$google) {
+      drupal_set_message($this->t('Social Auth Google not configured properly. Contact site administrator.'), 'error');
+      return $this->redirect('user.login');
+    }
 
-    // Gets user information.
-    $user = $this->googleManager->getUserInfo();
-    // If user information could be retrieved.
-    if ($user) {
-      return $this->userManager->authenticateUser($user->getEmail(), $user->getName(), $user->getId(), $user->getPicture());
+    $state = $this->dataHandler->get('oAuth2State');
+
+    // Retreives $_GET['state'].
+    $retrievedState = $this->request->getCurrentRequest()->query->get('state');
+    if (empty($retrievedState) || ($retrievedState !== $state)) {
+      $this->userManager->setSessionKeysToNullify(['oauth2state']);
```

Comment:
Change it by $this->userManager->nullifySessionKeys(). 

Since this means authentication fails, we should nullify session keys, so event handlers can check whether user was authenticated or not